File: forvo.py
from selenium import webdriver
import openpyxl
import selenium.common.exceptions
from selenium.webdriver.firefox.firefox_profile import FirefoxProfile
import time

from forvo_page import ForvoPage


# Test Setup
profile = FirefoxProfile()
profile.set_preference('browser.download.dir', '/Downloads')
profile.set_preference('browser.download.folderList', 2)
profile.set_preference(
    'browser.helperApps.neverAsk.saveToDisk', 'application/audio/mpeg')
browser = webdriver.Firefox(firefox_profile=profile)

# Go to Forvo
forvo_page = ForvoPage(driver=browser)
forvo_page.go('')
browser.maximize_window()
time.sleep(1)
forvo_page.agree_button.click()

# Log in to Forvo
forvo_page.username.input_text('')
forvo_page.password.input_text('')
forvo_page.login_button.click()


# reasearch before modularising
wb = openpyxl.load_workbook('4000.xlsx')
# print(type(wb))
# print(wb.sheetnames)
sheet = wb['Sheet1']

for i in range(2, 9):
    try:
        # forvo_page.keyword_button.click()
        website_link = sheet[f'C{i}'].value
        browser.get(website_link)
        time.sleep(3)
        print(i)
        keyword_pronounciation = forvo_page.pronounciation.text
        sheet[f'B{i}'].value = keyword_pronounciation
        wb.save('4000.xlsx')
    except Exception:
        sheet[f'B{i}'].value = 'no spelling'
        # continue
    try:
        # forvo_page.keyword_button.click()
        forvo_page.download_button.click()
        sheet[f'A{i}'].value = 'sound file downloaded'
    except Exception:
        sheet[f'A{i}'].value = 'no sound file'

wb.save('4000.xlsx')
print("operation finished! press any button to shut the browser down")
# forvo_page.download_button.click()

input()
browser.quit()

File: forvo.js
const {
  Builder,
  By,
  Key,
  until,
  size,
  actions,
} = require('selenium-webdriver');
const firefox = require('selenium-webdriver/firefox');
const options = new firefox.Options();

options.setPreference('browser.download.dir', '/Downloads');
options.setPreference('browser.download.folderList', 2);
options.setPreference(
  'browser.helperApps.neverAsk.saveToDisk',
  'application/audio/mpeg'
);

const driver = new Builder()
  .forBrowser('firefox')
  .setFirefoxOptions(options)
  .build();

// (async function() {
// await driver.get("https://forvo.com/download/mp3/any/en/2162053");
// await driver.quit();
// })();

var details = ['', ''];

var keywords = [
  'a',
  'abandon',
  'ability',
  'able',
  'abortion',
  'about',
  'above',
  'abroad',
  'absence',
  'absolute',
  'absolutely',
  'absorb',
  'abuse',
  'academic',
  'accept',
  'access',
  'accident',
  'accompany',
  'accomplish',
  'according',
  'account',
  'accurate',
  'accuse',
  'achieve',
  'achievement',
  'acid',
  'acknowledge',
  'acquire',
  'across',
  'act',
  'action',
  'active',
  'activist',
  'activity',
  'actor',
  'actress',
  'actual',
  'actually',
  'ad',
  'adapt',
  'add',
  'addition',
  'additional',
  'address',
  'adequate',
  'adjust',
  'adjustment',
  'administration',
];

proList = [];

notCatched = [];

(async function () {
  //log in into the website
  await login();

  //navigate to website
  await navigate();

  async function login() {
    await driver.get('https://forvo.com/login/');
    // let logIn = await driver.findElement(By.css('.login'));
    // await logIn.click();
    await driver.sleep(2000);
    let username = await driver.findElement(By.css('#login'));
    await username.click();
    await driver.sleep(200);
    await username.sendKeys(details[0]);
    let password = await driver.findElement(By.css('#password'));
    await password.click();
    await driver.sleep(200);
    await password.sendKeys(details[1]);
    await driver.findElement(By.css('.button')).click();
  }

  async function navigate() {
    for (i = 0; i < keywords.length; i++) {
      await driver.get('https://forvo.com/search/' + keywords[i] + '/en_uk/');
      await driver.sleep(1000);

      // download the file with pronounciation to Downloads folder and push spelling to array proList
      await clickDownloadFileAndSpelling();
      console.log(proList);
    }
    console.log(proList, notCatched);
    console.log(proList.length, keywords.length, notCatched.length);
  }

  async function clickDownloadFileAndSpelling() {
    try {
      let word = await driver.findElement(
        By.css(
          'ul.word-play-list-icon-size-l:nth-child(1) > li:nth-child(1) > a:nth-child(2)'
        )
      );
      await word.click();
      let pronounciation = await driver.findElement(By.css('.phonetics'));
      await pronounciation.getText().then(function (txt) {
        proList.push(` ${txt} `);
      });
      let sweet = await driver.findElement(
        By.xpath("//*[@id='en_uk']/following-sibling::*/div/div/p[3]")
      );
      await sweet.click();
    } catch {
      proList.push('NOT FOUND');
      notCatched.push(keywords[i]);
      console.log(notCatched);
      console.log(i);
    }
  }

  await driver.quit();
})();
